rfcs/20240315-algorithm-support.md

```diff
@@ -0,0 +1,165 @@
+# [RFC] Add algorithm to dot_general's parameters in the StableHLO specification
+
+Status: Review<br/>
+Initial version: 03/15/2024<br/>
+Last updated: 03/15/2024<br/>
+Discussion thread: [GitHub](https://github.com/openxla/stablehlo/pull/2096)
+
+## Motivation
+
+We would like to allow explicitly selecting the algorithm used for individual
+`dot_general` instructions, enforcing a well-defined numeric precision. We think
+that the current precision values (`DEFAULT`, `HIGH`, `HIGHEST`) are not
+specific enough for this, because they mean different things between hardware
+platforms and even between GPU types.
+
+We also want to make the API flexible enough to support additional algorithms
+that might be faster on certain hardware
+([6xBF16](https://arxiv.org/pdf/1904.06376.pdf), 3xTF32, etc) that can't fit
+into the limited precision_config values of `DEFAULT`, `HIGH` and `HIGHEST`. For
+example "6xBF16" provides similar precision to F32, but about 2x performance on
+some GPUs.
+
+When an algorithm is not supported by an accelerator, the program should fail
+instead of implicitly falling back to another. It should be the responsibility
+of higher level frameworks such as JAX, to define the required algorithm for
+each accelerator that is used.
+
+## Proposed Specification change
+
+### dot_general
+
+#### Semantics
+
+*Add these lines:*
+
+`algorithm` defines the main properties of the algorithm used to implement the
+dot operation, which also defines the precision. If the algorithm value is set
+(`!= UNSET`), then the `precision_config` is ignored.
```

Comment:
Does that also mean that if the algorithm is not set then precision_config will take effect? If yes, can we clarify that as well.

Reply:
Added the requested clarification.